service/collector.go

```diff
@@ -158,7 +158,8 @@ func (col *Collector) setupConfigurationComponents(ctx context.Context) error {
 // Run starts the collector according to the given configuration, and waits for it to complete.
 // Consecutive calls to Run are not allowed, Run shouldn't be called once a collector is shut down.
 func (col *Collector) Run(ctx context.Context) error {
-	if err := col.setupConfigurationComponents(ctx); err != nil {
+	_, ok := <-col.shutdownChan
```

Comment:
If `Shutdown` is not called before `Run` this will block indefinitely causing run to never start.

I think we can add a check like this at the start:

```go
state := col.GetState()
if state == Closing || state == Closed {
  return errors.New("collector has already been shutdown and cannot be run again")
}
```

It might be worth saving that error off into a variable for easy of use and testing.

Reply:
I really appreciate your help and guidance towards helping me complete this issue, I guess I am still a bit new to GO and GO's concurrency is uncharted waters for me, glad for this issue that has served to help me learn it practically. So while coming up with a solution, I looked at the test case you created [here]((https://github.com/open-telemetry/opentelemetry-collector/pull/4878/files#diff-39cb0a393ab609b58e27ab97198f7fdd885a5369e325d25b32879b22d6e19019) and I created a small snippet of code to simulate that. When I ran it even without "shutdown" being called before, it worked so I thought it was a good solution:

![image](https://user-images.githubusercontent.com/59079323/195477871-49acb04d-7a26-4977-8d26-084b1260671c.png)
I thought since there is an equivalent concurrent reading and writing of the channel in the main function and go routine, there would be no blocking. I guess I might have missed something.

It did block if shut is not called right after run, but I found that was the same case if I did not have that extra line checking if the  channel was closed or not:

![image](https://user-images.githubusercontent.com/59079323/195479024-93d6052d-4c7a-4649-adc0-a1e8419b1386.png)


Your solution looks really good but I guess I did not not know the collection's state would be set to closing when the `ShutDown` function was called, as I could not find it here:

![image](https://user-images.githubusercontent.com/59079323/195478542-150919bd-b45a-4d28-90f5-834d5e0d6d35.png)

Reply:
>Your solution looks really good but I guess I did not not know the collection's state would be set to closing when the ShutDown function was called, as I could not find it here:

It's a bit of a chain reaction. Closing the `shutdownChan` gets picked up in `Run` [here](https://github.com/open-telemetry/opentelemetry-collector/blob/main/service/collector.go#L195). It then calls `col.shutdown` which sets the state to `Closing` [here](https://github.com/open-telemetry/opentelemetry-collector/blob/main/service/collector.go#L209).

As for learning go concurrency if you haven't yet checking out [tour of go](https://go.dev/tour/concurrency/1) is a good starting place. Also, just playing around with it is a good way to learn. In the example above you gave this line will block until something is pushed onto the channel or the channel is closed.

```go
_, ok := <-col.shutdownChan
```

If the channel is closed `ok` will be false. If something is pushed the channel, before closing, `ok` will be true indicating the channel is still active and can be listened on again.

Reply:
Thank you so much. I find that there are two shutdown functions.One with a capital 's' and the other with a small 's' The one I saw in your test and on the screenshot that I shared above is the one with a capital 's'. So when collector.ShutDown is called it does not cause any chain reaction as it is not same as collector.shutdown which is used inside in Run() function.

Thanks for the great tips on how to get started with concurrency. 

`_, ok := <-col.shutdownChan`
This line reads from the col channel  and waits till there something os pushed to the channel before it unblocks. Can we not say the same about the select statement in Run() function? As it does not have a 'default block'. I find that if you call The Run() function without the ShutDown function without the ShutDown function, before or after the Run(), we would get an error
So I guess adding the line of code that I added would not make the function behave differently

Reply:
>So when collector.ShutDown is called it does not cause any chain reaction as it is not same as collector.shutdown which is used inside in Run() function.

So here is the chain reaction that causes the collector to stop once `Shutdown` is called:
1. [Shutdown is called](https://github.com/open-telemetry/opentelemetry-collector/blob/main/service/collector.go#L114)
2. If the collector is running it [closes the shutdownChan](https://github.com/open-telemetry/opentelemetry-collector/blob/main/service/collector.go#L121)
3. In `Run` the closed channel triggers [this](https://github.com/open-telemetry/opentelemetry-collector/blob/main/service/collector.go#L195) case in the select and breaks out of the main loop.
4. At the end of `Run` [shutdown is called](https://github.com/open-telemetry/opentelemetry-collector/blob/main/service/collector.go#L205) and `Run` returns the result of `shutdown`.

> This line reads from the col channel and waits till something is pushed to the channel, causing a block. Can we not say the same about the select statement in Run() function?

This is true. The issue #4946 that you're working on is to make this happen immediately inside `Run`. Currently `Run` has some setup logic [here](https://github.com/open-telemetry/opentelemetry-collector/blob/main/service/collector.go#L161-L164) which calls [setupConfigurationComponents](https://github.com/open-telemetry/opentelemetry-collector/blob/f24a6ac30fa027d3e373f6280f5b4bb4b4626587/service/collector.go#L127-L156). There are a lot of things that happen in this that can all be skipped by an early detection that we've already been shut down.

Reply:
Thanks a lot for your patience with me. This chain of event only happens if Run is called before ShutDown. I thought we were looking at a scenario when Shutdown is called before Run. As per your statement:

![image](https://user-images.githubusercontent.com/59079323/195665615-e440e63e-4bbe-4178-bdea-79bbd0db2f2a.png)

So the way that I am looking at it is if ShutDown is called before Run, it would not be able to receive any message, as the blocking event awaiting receipt of message from shutDownchan is not currently on (i.e. Run has not been called yet) but I guess that it is not how it works. That was my initial thought process :

![image](https://user-images.githubusercontent.com/59079323/195672900-49c8c2e0-0d99-43f5-abb2-3449397577f2.png)

Thanks for the clarity.

Then I guess even though the select statement causes a block, it makes that happen in the right stage of execution as opposed to where my line of code would have caused that